cmd/atlas/internal/cmdapi/diff.go

```diff
@@ -17,85 +24,295 @@ import (
+	if err != nil {
+		return err
+	}
+	defer from.Close()
+	to, err := stateReader(ctx, &stateReaderConfig{urls: []string{flags.toURL}, dev: c})
+	if err != nil {
+		return err
+	}
+	defer to.Close()
+	if c == nil {
+		var ok bool
+		c, ok = from.Closer.(*sqlclient.Client)
+		if !ok {
+			c, ok = to.Closer.(*sqlclient.Client)
+			if !ok {
+				// There is no driver to diff. We need a dev-database.
+				cobra.CheckErr(errors.New(
+					"neither --from nor --to provides a database connection, please provide one by passing in a --dev-url",
+				))
+			}
+		}
+	}
+	current, err := from.ReadState(ctx)
+	if err != nil {
+		return err
+	}
+	desired, err := to.ReadState(ctx)
+	if err != nil {
+		return err
+	}
 	var diff []schema.Change
 	switch {
-	case fromS == "" && toS == "":
-		// compare realm.
-		fromRealm, err := fromC.InspectRealm(ctx, nil)
-		cobra.CheckErr(err)
-		toRealm, err := toC.InspectRealm(ctx, nil)
-		cobra.CheckErr(err)
-		diff, err = toC.RealmDiff(fromRealm, toRealm)
-		cobra.CheckErr(err)
-	case fromS == "":
-		cobra.CheckErr(fmt.Errorf("cannot diff schema %q with a database connection", fromS))
-	case toS == "":
-		cobra.CheckErr(fmt.Errorf("cannot diff database connection with a schema %q", toS))
+	// compare realm
+	case from.Schema == "" && to.Schema == "":
+		diff, err = c.RealmDiff(current, desired)
+		if err != nil {
+			return err
+		}
+	case from.Schema == "":
+		cobra.CheckErr(fmt.Errorf("cannot diff schema %q with a database connection", from.Schema))
+	case to.Schema == "":
+		cobra.CheckErr(fmt.Errorf("cannot diff database connection with a schema %q", to.Schema))
 	default:
-		// compare schemas.
-		fromSchema, err := fromC.InspectSchema(ctx, fromS, nil)
-		cobra.CheckErr(err)
-		toSchema, err := toC.InspectSchema(ctx, toS, nil)
-		cobra.CheckErr(err)
 		// SchemaDiff checks for name equality which is irrelevant in the case
-		// the user wants to compare their contents, if the names are different
-		// we reset them to allow the comparison.
-		if fromS != toS {
-			toSchema.Name = ""
-			fromSchema.Name = ""
+		// the user wants to compare their contents, reset them to allow the comparison.
+		current.Schemas[0].Name, desired.Schemas[0].Name = "", ""
+		diff, err = c.SchemaDiff(current.Schemas[0], desired.Schemas[0])
+		if err != nil {
+			return err
 		}
-		diff, err = toC.SchemaDiff(fromSchema, toSchema)
-		cobra.CheckErr(err)
 	}
-	p, err := toC.PlanChanges(ctx, "plan", diff)
-	cobra.CheckErr(err)
+	p, err := c.PlanChanges(ctx, "plan", diff)
+	if err != nil {
+		return err
+	}
 	if len(p.Changes) == 0 {
 		cmd.Println("Schemas are synced, no changes to be made.")
-		return
+		return nil
 	}
 	for _, c := range p.Changes {
 		if c.Comment != "" {
 			cmd.Println("--", strings.ToUpper(c.Comment[:1])+c.Comment[1:])
 		}
 		cmd.Println(c.Cmd)
 	}
+	return nil
+}
+
+type (
+	// stateReadCloser is a migrate.StateReader with an optional io.Closer.
+	stateReadCloser struct {
```

Comment:
The `target` struct in `migrate-.go` provides same behavior and will be merged with this one once the migrate commands are touched.